Cache MEF catalog in servicehub process

ToddGrun opened the pull request:
Cache the MEF catalog created in the roslyn code analysis process.

On my dev machine, the creation of the MEF catalog in our CA process takes about 500 ms. With this change, runs that were able to load from the cache saw that time reduced to about 150 ms.

The high-level change here is to piggy back on the work that Joey did a while back to add MEF catalog caching to our VSCode language server. This PR just moves that code down to a layer where our service hub code can reference, and generalized it a bit to fit into what the service hub code can provide. 

One slight additional change in this PR was to add a new remote service to handle passing over the proper location to store the mef cache. I didn't feel comfortable fitting this into the previous first message sent across (IRemoteProcessTelemetryService), so I created IRemoteInitializationService to better fit what I wanted. This led to creating a separate type to handle the export provider creation (RemoteExportProvider) instead of tying this up into RemoteWorkspaceManager.

dibarbet commented on src/Workspaces/Remote/ServiceHub/Host/RemoteExportProvider.cs:
```diff
@@ -0,0 +1,69 @@
+        var cacheDirectory = Path.Combine(localSettingsDirectory, "Roslyn", "RemoteHost", "Cache");
+        var catalogPrefix = "RoslynRemoteHost";
+
+        s_instance = await ExportProviderBuilder.CreateExportProviderAsync(assemblyPaths, resolver, cacheDirectory, catalogPrefix, performCleanup: true, loggerFactory: null, cancellationToken).ConfigureAwait(false);
```
should work, but would be good to verify - in vscode extension version updates change the actual extension folder, so the cache can never be re-used across updates.  But in VS I assume that is not true.  We do use the assembly write time as part of the key so I think it should be fine, but not as confident

ToddGrun replied:
I'm not clear on the concern. It looks like the export provider builder is based completely on last file write times (in addition to the full path). If either of those have changed, we should get a new cache. Is there a concern where those checks aren't sufficient?

ToddGrun commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -151,44 +136,52 @@ static string ComputeAssemblyHash(ImmutableArray<string> assemblyPaths)
+                if (performCleanup)
+                {
+                    // Delete any existing cached files.
+                    foreach (var fileInfo in directoryInfo.EnumerateFiles($"*{CatalogSuffix}"))
```
MultipleMefCompositionsAreCached was failing until I added this check

dibarbet replied:
cc @JoeRobich do you remember why we needed multiple?

JoeRobich replied:
"need" might be a strong word here. 

Since there are projects that the CDK doesn't support, our users may have a valid reason to at times use C# on its own in addition to C# w/ CDK. The caches aren't terribly big and are cleaned up on each upgrade of the extension because we store them in the extension folder.

dibarbet replied:
Ah right - cdk vs non CDK, as well as different extensions in different profiles

ToddGrun replied:
Per our offline conversation, this cache is VS instance and hive aware, and thus caches for different versions or hives should not affect each other.

ToddGrun commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -151,44 +136,52 @@ static string ComputeAssemblyHash(ImmutableArray<string> assemblyPaths)
 
+#if NET
             File.Move(tempFilePath, compositionCacheFile, overwrite: true);
+#else
```
This project is built against netstandard2.0

ToddGrun commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -151,44 +136,52 @@ static string ComputeAssemblyHash(ImmutableArray<string> assemblyPaths)
     }
 
-    private static async Task WriteCompositionCacheAsync(string compositionCacheFile, CompositionConfiguration config, ILogger logger)
+    private static async Task WriteCompositionCacheAsync(string compositionCacheFile, CompositionConfiguration config, bool performCleanup, ILogger? logger, CancellationToken cancellationToken)
```
I *think* we'd be ok, as if the cache state is corrupsed, then the cachedComposition.LoadExportProviderFactoryAsync will throw and we'll just recreate the cache.

dibarbet commented on src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LanguageServerExportProviderBuilder.cs:
```diff
@@ -0,0 +1,67 @@
+        // Also add the ExtensionAssemblyManager so it will be available for the rest of the composition.
+        exportProvider.GetExportedValue<ExtensionAssemblyManagerMefProvider>().SetMefExtensionAssemblyManager(extensionManager);
+
+        var exportProviderBuilderLogger = loggerFactory.CreateLogger<ExportProviderBuilder>();
```
seems unused

ToddGrun replied:
Yep, probably a remnant of when I was refactoring the logging code. Nice catch!

dibarbet reviewed (approved):
Someone else should likely review the servicehub initialization part

ToddGrun commented on src/Workspaces/Remote/ServiceHub/Host/RemoteExportProvider.cs:
```diff
@@ -0,0 +1,73 @@
+            Resolver: new Resolver(SimpleAssemblyLoader.Instance),
+            CacheDirectory: Path.Combine(localSettingsDirectory, "Roslyn", "RemoteHost", "Cache"),
+            CatalogPrefix: "RoslynRemoteHost",
+            ExpectedErrorParts: ["PythiaSignatureHelpProvider", "VSTypeScriptAnalyzerService", "RazorTestLanguageServerFactory", "CodeFixService"],
```
>"CodeFixService" [](http://example.com/codeflow?start=129&length=16)

@CyrusNajmabadi -- any idea why this type has started not resolving in the CA process MEF composition?

jasonmalinowski replied:
Is that some fallout of the recent removal of EditorFeatures? That was rolled back, but it's still suspicious.

ToddGrun replied:
I merged main and still need this type still errors. 
```
Microsoft.CodeAnalysis.CodeFixes.CodeFixService.ctor(diagnosticAnalyzerService): expected exactly 1 export matching constraints:
    Contract name: Microsoft.CodeAnalysis.Diagnostics.IDiagnosticAnalyzerService
    TypeIdentityName: Microsoft.CodeAnalysis.Diagnostics.IDiagnosticAnalyzerService
but found 0.
```

CyrusNajmabadi replied:
my pr hasn't merged in yet.

ToddGrun replied:
OK, I tested with both our branches combined and it gets rid of the errors for VSTypeScriptAnalyzerService and CodeFixService

ToddGrun replied:
Here are the errors for those two:
```
Microsoft.CodeAnalysis.ExternalAccess.Razor.RazorTestLanguageServerFactory.ctor(languageServerFactory): expected exactly 1 export matching constraints:
    Contract name: Microsoft.CodeAnalysis.LanguageServer.ILanguageServerFactory
    TypeIdentityName: Microsoft.CodeAnalysis.LanguageServer.ILanguageServerFactory
but found 0.

Microsoft.CodeAnalysis.ExternalAccess.Pythia.PythiaSignatureHelpProvider.ctor(implementation): expected exactly 1 export matching constraints:
    Contract name: Microsoft.CodeAnalysis.ExternalAccess.Pythia.Api.IPythiaSignatureHelpProviderImplementation
    TypeIdentityName: Microsoft.CodeAnalysis.ExternalAccess.Pythia.Api.IPythiaSignatureHelpProviderImplementation
but found 0.
```

ToddGrun commented:
@jasonmalinowski -- ptal

---
In reply to: [2761634455](https://github.com/dotnet/roslyn/pull/78122#issuecomment-2761634455) [](http://example.com/codeflow?ancestors=2761634455)

CyrusNajmabadi commented on src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LanguageServerExportProviderBuilder.cs:
```diff
@@ -0,0 +1,65 @@
+        var baseDirectory = AppContext.BaseDirectory;
+
+        // Load any Roslyn assemblies from the extension directory
+        var assemblyPaths = Directory.EnumerateFiles(baseDirectory, "Microsoft.CodeAnalysis*.dll");
```
do we need do any catching of IO issues?

ToddGrun replied:
This code probably avoided that complexity originally (and thus why I didn't change that) as I'm guessing a partial export provider built up is not really any better than no export provider.

jasonmalinowski replied:
Yep -- better to crash right away than crash with obscure MEF issues that will take a long time to sort out.

CyrusNajmabadi commented on src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LanguageServerExportProviderBuilder.cs:
```diff
@@ -0,0 +1,65 @@
+
+        // Load any Roslyn assemblies from the extension directory
+        var assemblyPaths = Directory.EnumerateFiles(baseDirectory, "Microsoft.CodeAnalysis*.dll");
+        assemblyPaths = assemblyPaths.Concat(Directory.EnumerateFiles(baseDirectory, "Microsoft.ServiceHub*.dll"));
```
are tehse all IEnumerables we're lazily concatenating?

CyrusNajmabadi commented on src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LanguageServerExportProviderBuilder.cs:
```diff
@@ -0,0 +1,65 @@
+        }
+
+        // Add the extension assemblies to the MEF catalog.
+        assemblyPaths = assemblyPaths.Concat(extensionManager.ExtensionAssemblyPaths);
```
consider an actual array-builder, just to have the work be clear where it happens.

ToddGrun replied:
switched in next commit

CyrusNajmabadi commented on src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LanguageServerExportProviderBuilder.cs:
```diff
@@ -0,0 +1,65 @@
+            CatalogPrefix: "c#-languageserver",
+            ExpectedErrorParts: ["CSharpMapCodeService", "PythiaSignatureHelpProvider", "CopilotSemanticSearchQueryExecutor"],
+            PerformCleanup: false,
+            LogError: text => logger.LogError(text),
+            LogTrace: text => logger.LogTrace(text));
```
```suggestion
            LogError: logger.LogError,
            LogTrace: logger.LogTrace);
```

ToddGrun replied:
I originally tried that and got:

![image](https://github.com/user-attachments/assets/87316c01-7605-4e65-891f-08dfbe9d0892)

jasonmalinowski replied:
Yep those are extension methods, so that won't easily work.

ToddGrun replied:
I added a reference to Microsoft.Extensions.Logging in the first commit and then shied away from it as it yielded build time warning about needing to add those files to the list we ship in our SH process. So, I stopped that.

dibarbet replied:
Yeah I suspect it will cause an RPS regression (though not 100% sure).  We could do a PR validation on it

CyrusNajmabadi commented on src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Services/ExtensionAssemblyManager.cs:
```diff
@@ -7,6 +7,7 @@
 using System.Runtime.Loader;
 using Microsoft.CodeAnalysis.LanguageServer.StarredSuggestions;
 using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Remote;
```
remove?

ToddGrun replied:
Needed tor ExportProviderBuilder

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Remote/VisualStudioRemoteHostClientProvider.cs:
```diff
@@ -44,6 +48,7 @@ internal sealed class Factory : IWorkspaceServiceFactory
         public Factory(
             VisualStudioWorkspace vsWorkspace,
             IVsService<SVsBrokeredServiceContainer, IBrokeredServiceContainer> brokeredServiceContainer,
+            SVsServiceProvider serviceProvider,
```
would it make sense to take an IVsService like above, since you just use the iServiceProvider side of htis?

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Telemetry/IRemoteProcessTelemetryService.cs:
```diff
@@ -21,11 +20,4 @@ internal interface IRemoteProcessTelemetryService
-    /// Called as soon as the remote process is created but can't guarantee that solution entities (projects, documents, syntax trees) have not been created beforehand.
-    /// </summary>
-    /// <returns>Process ID of the remote process.</returns>
-    ValueTask<int> InitializeAsync(WorkspaceConfigurationOptions options, CancellationToken cancellationToken);
```
ah, so this moved to teh new service?

CyrusNajmabadi commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -0,0 +1,210 @@
+
+namespace Microsoft.CodeAnalysis.Remote;
+
+internal sealed class ExportProviderBuilder
```
nit: considre calling RemoteExportProviderBuilder.  THe first part helps prime me that this is not a general thing, but exists for the remote side.  up to you.

ToddGrun replied:
Kept the original name as this is used by the language server too

CyrusNajmabadi commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -0,0 +1,210 @@
+        // Try to load a cached composition.
+        try
+        {
+            if (File.Exists(compositionCacheFile))
```
what ensures that just because it exists it is ok to load and wont' be stale?  might be worth docing (here, or somewhere0.

ToddGrun replied:
Added a doc comment to GetCompositionCacheFilePath

CyrusNajmabadi commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -0,0 +1,210 @@
+
+    private static string GetCompositionCacheFilePath(string cacheDirectory, string catalogPrefix, ImmutableArray<string> assemblyPaths)
+    {
+        return Path.Combine(cacheDirectory, $"{catalogPrefix}.{ComputeAssemblyHash(assemblyPaths)}{CatalogSuffix}");
```
interesting.  so we effectively 'leak' the composition caches over time?  not complainign, just trying to understand.

ToddGrun replied:
There shouldn't be any leaks of caches, the path includes the VS instance and hive, and any mef compositions under that path that don't match the currently hashed cache file name should be deleted.

CyrusNajmabadi commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -0,0 +1,210 @@
+
+            var hash = XxHash128.Hash(Encoding.UTF8.GetBytes(hashContents.ToString()));
+            // Convert to filename safe base64 string.
+            return Convert.ToBase64String(hash).Replace('+', '-').Replace('/', '_').TrimEnd('=');
```
just use `Checksum.Create(hashContents.ToString).ToString()`

CyrusNajmabadi commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -0,0 +1,210 @@
+    {
+        try
+        {
+            await Task.Yield();
```
```suggestion
            await Task.Yield().ConfigureAwait(false);
```

CyrusNajmabadi commented on src/Workspaces/Remote/ServiceHub/Host/RemoteExportProvider.cs:
```diff
@@ -0,0 +1,73 @@
+            .Add(typeof(IExtensionWorkspaceMessageHandler<,>).Assembly);
+
+    private static ExportProvider? s_instance;
+    internal static ExportProvider ExportProvider
+    {
+        get
+        {
+            Contract.ThrowIfNull(s_instance, "Default export provider not initialized. Call InitializeAsync first.");
+            return s_instance;
+        }
+    }
```
```suggestion
    internal static ExportProvider ExportProvider
        => s_instance ?? throw new InvalidOperationException("Default export provider not initialized. Call InitializeAsync first.");
```

CyrusNajmabadi commented on src/Workspaces/Remote/ServiceHub/Host/RemoteExportProvider.cs:
```diff
@@ -0,0 +1,73 @@
+            ExpectedErrorParts: ["PythiaSignatureHelpProvider", "VSTypeScriptAnalyzerService", "RazorTestLanguageServerFactory", "CodeFixService"],
+            PerformCleanup: true,
+            LogError: static _ => { },
+            LogTrace: static _ => { });
```
why no logging?  just no way to support it?  it seems like this call could be passed an RPC callback to send this info.

ToddGrun replied:
I didn't see an obvious way to pipe things through. Logging this early in the remote host initialization process isn't trivial.

jasonmalinowski replied:
Can we at least write to a string builder and send the string back from InitializeAsync? And propagate that somewhere if the error string is non-empty?

ToddGrun replied:
Went ahead and pushed back any error messages occurred during the remote initialize call.

CyrusNajmabadi commented on src/Workspaces/Remote/ServiceHub/Host/RemoteExportProvider.cs:
```diff
@@ -0,0 +1,73 @@
+#pragma warning disable SYSLIB0044 // https://github.com/dotnet/roslyn/issues/71510
+                assemblyName.CodeBase = codeBasePath;
+#pragma warning restore SYSLIB0044
+            }
```
docs...

CyrusNajmabadi commented on src/Workspaces/Remote/ServiceHub/Services/Initialization/RemoteInitializationService.cs:
```diff
@@ -0,0 +1,38 @@
+    /// </summary>
+    public async ValueTask<int> InitializeAsync(WorkspaceConfigurationOptions options, string localSettingsDirectory, CancellationToken cancellationToken)
+    {
+        await RemoteExportProvider.InitializeAsync(localSettingsDirectory, cancellationToken).ConfigureAwait(false);
```
move into RunServiceAsync?

ToddGrun replied:
Adding a comment why that's a bad idea.

jasonmalinowski replied:
It might be worth expanding on that comment if the reason it's broken is because something in RunServiceAsync depends on it...

ToddGrun commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -0,0 +1,210 @@
+                hashContents.Append(assemblyPath);
+                // Include the last write time in the hash so that newer assemblies written
+                // to the same location cause the composition to be rebuilt.
+                hashContents.Append(File.GetLastWriteTimeUtc(assemblyPath).ToString("F"));
```
I'm hesitant to change this to checksum the file contents. At the very worst, this gets the wrong last file time and we just recreate the cache.

jasonmalinowski commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -0,0 +1,220 @@
+
+namespace Microsoft.CodeAnalysis.Remote;
+
+internal sealed class ExportProviderBuilder
```
```suggestion
internal static class ExportProviderBuilder
```

jasonmalinowski commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -0,0 +1,220 @@
+{
+    private const string CatalogSuffix = ".mef-composition";
+
+    // For testing purposes, track the last cache write task.
+    private static Task? s_cacheWriteTask;
```
Can we put this bit back in the language server layer, if it's only being used there?

jasonmalinowski commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -0,0 +1,220 @@
+        static string ComputeAssemblyHash(ImmutableArray<string> assemblyPaths)
+        {
+            // Ensure AssemblyPaths are always in the same order.
+            assemblyPaths = assemblyPaths.Sort();
```
Should this be specifying some comparer? If it's accidentally going to inherit a culture-specific comparer then it might not be as stable as you might want.

jasonmalinowski commented on src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LanguageServerExportProviderBuilder.cs:
```diff
@@ -0,0 +1,63 @@
+
+namespace Microsoft.CodeAnalysis.LanguageServer;
+
+internal sealed class LanguageServerExportProviderBuilder
```
```suggestion
internal static class LanguageServerExportProviderBuilder
```

jasonmalinowski commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -0,0 +1,220 @@
+                    // Delete any existing cached files.
+                    foreach (var fileInfo in directoryInfo.EnumerateFiles($"*{CatalogSuffix}"))
+                    {
+                        fileInfo.Delete();
```
I assume you want a try/catch around this -- if this fails to cleanup you probably just want to just leave the other one there and assume it'll get cleaned up the next time.

ToddGrun replied:
makes sense

jasonmalinowski commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -0,0 +1,220 @@
+            }
+
+            CachedComposition cachedComposition = new();
+            var tempFilePath = Path.Combine(Path.GetTempPath(), Path.GetTempFileName());
```
This should probably create the temporary file in the same directory for a few reasons:

1. Security: on Windows %TEMP% is isolated from users; that's not necessarily the case on Linux. This at least creates a small information leak in that the user's installed extensions are present in the MEF cache.
2. Reliability: it makes a stronger guarantee that that File.Move() can't corrupt stuff. If it's on the same volume/filesystem, that becomes a bit of file system recordkeeping and that's it. And you can use File.Replace() if needed. If they're on different file/systems, then the move is under the cover copying bits, and a badly timed power outage means your cache file might get corrupted.

ToddGrun replied:
makes sense

jasonmalinowski commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -0,0 +1,220 @@
+#if NET
+            File.Move(tempFilePath, compositionCacheFile, overwrite: true);
+#else
+            // On .NET Framework, File.Move doesn't support overwriting the destination file. Use File.Delete first
+            // to ensure the destination file is removed before moving the temp file. File.Delete will not throw if
+            // the file doesn't exist.
+            File.Delete(compositionCacheFile);
```
Why would the destination file exist? Woudln't we have then used that cache in that case?

ToddGrun replied:
It really shouldn't, will just switch to the File.Move that exists in both TF that doesn't take in the overwrite flag

jasonmalinowski commented on src/Workspaces/Remote/Core/IRemoteInitializationService.cs:
```diff
@@ -0,0 +1,19 @@
+{
+    /// <summary>
+    /// Initializes values including <see cref="WorkspaceConfigurationOptions"/> for the process.
+    /// Called as soon as the remote process is created but can't guarantee that solution entities (projects, documents, syntax trees) have not been created beforehand.
```
> can't guarantee that solution entities (projects, documents, syntax trees) have not been created beforehand

Is that possible, since any of the entities we would have created would have needed MEF to be created first? I think this comment you copied/pasted and its stale now.

ToddGrun replied:
Will remove, it didn't make sense to me, but I don't have a lot of context in this area.

jasonmalinowski commented on src/Workspaces/Remote/ServiceHub/Host/RemoteExportProvider.cs:
```diff
@@ -0,0 +1,69 @@
+{
+    internal static readonly ImmutableArray<Assembly> RemoteHostAssemblies =
+        MefHostServices.DefaultAssemblies
+            .Add(typeof(AspNetCoreEmbeddedLanguageClassifier).Assembly)
+            .Add(typeof(BrokeredServiceBase).Assembly)
+            .Add(typeof(IRazorLanguageServerTarget).Assembly)
+            .Add(typeof(RemoteWorkspacesResources).Assembly)
+            .Add(typeof(IExtensionWorkspaceMessageHandler<,>).Assembly);
```
Consider adding comments for each of these for which assembly name that is.

jasonmalinowski commented on src/Workspaces/Remote/ServiceHub/Host/RemoteExportProvider.cs:
```diff
@@ -0,0 +1,69 @@
+
+    private static ExportProvider? s_instance;
+    internal static ExportProvider ExportProvider
+        => s_instance ?? throw new InvalidOperationException("Default export provider not initialized. Call InitializeAsync first.");
```
```suggestion
        => s_instance ?? throw new InvalidOperationException($"Default export provider not initialized. Call {nameof(InitializeAsync)} first.");
```

jasonmalinowski commented:
@ToddGrun Reviewed -- many comments are on things you moved so they're not necessarily your fault, but might be worth addressing if we can anyways.

ToddGrun commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -0,0 +1,220 @@
+                await cachedComposition.SaveAsync(config, cacheStream, cancellationToken).ConfigureAwait(false);
+            }
+
+#if NET
+            File.Move(tempFilePath, compositionCacheFile, overwrite: true);
+#else
+            // On .NET Framework, File.Move doesn't support overwriting the destination file. Use File.Delete first
+            // to ensure the destination file is removed before moving the temp file. File.Delete will not throw if
+            // the file doesn't exist.
+            File.Delete(compositionCacheFile);
+            File.Move(tempFilePath, compositionCacheFile);
+#endif
```
The documentation is wonky, but that's what I see when I try to use it.

ToddGrun commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -0,0 +1,210 @@
+        ThrowOnUnexpectedErrors(config, catalog, args.ExpectedErrorParts, args.LogError);
+
+        // Try to cache the composition.
+        _cacheWriteTask = WriteCompositionCacheAsync(compositionCacheFile, config, args.PerformCleanup, args.LogError, cancellationToken).ReportNonFatalErrorAsync();
```
Let me know if there is still a concern here.

ToddGrun commented:
> Reviewed -- many comments are on things you moved so they're not necessarily your fault, but might be worth addressing if we can anyways.

@jasonmalinowski -- I think I addressed most of the concerns, can you take another peek and make sure the changes match your expectations?

jasonmalinowski commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -50,65 +47,70 @@ public static async Task<ExportProvider> CreateExportProviderAsync(
         {
             // Log the error, and move on to recover by recreating the MEF composition.
-            args.LogError($"Loading cached MEF composition failed: {ex}");
+            LogError($"Loading cached MEF composition failed: {ex}");
```
Consider letting LogError take an exception, as then we can log it to telemetry more nicely.

ToddGrun replied:
I think I'd like to avoid making any more decisions on how to expose this error information in this PR. I don't know if the additional information that can be gleaned from an exception is of much value in this situation.

jasonmalinowski commented on src/Workspaces/Remote/Core/ServiceHubRemoteHostClient.cs:
```diff
@@ -73,15 +73,18 @@ public static async Task<RemoteHostClient> CreateAsync(
             var workspaceConfigurationService = services.GetRequiredService<IWorkspaceConfigurationService>();
 
-            var remoteProcessId = await client.TryInvokeAsync<IRemoteInitializationService, int>(
+            var remoteProcessIdAnderrorMessage = await client.TryInvokeAsync<IRemoteInitializationService, (int ProcessId, string? ErrorMessage)>(
```
```suggestion
            var remoteProcessIdAndErrorMessage = await client.TryInvokeAsync<IRemoteInitializationService, (int ProcessId, string? ErrorMessage)>(
```

jasonmalinowski reviewed (approved):
I'm signing off, but CI is less happy with it. :smile:

ToddGrun commented:
Failing now due razor ea causing MEF load failures after merging in the latest main. Pinged Andrew.

1) RazorDynamicFileInfoProviderWrapper errors on construction (this I can work around easily)
2) ReflectionTypeLoadException occurs during composing this assembly due to the following:
 
ReflectionTypeLoadException while enumerating types in assembly "c:\users\toddgrun\appdata\local\microsoft\visualstudio\18.0_1d825a96roslyndev\extensions\microsoft\roslyn services for .net servicehub host\42.42.42.4242424\Microsoft.CodeAnalysis.ExternalAccess.Razor.dll". Results will be incomplete.
 
InnerException: 
Unable to load one or more of the requested types.
Could not load file or assembly 'Microsoft.VisualStudio.LanguageServer.Client, Version=17.13.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.

ToddGrun commented:
Test insertion doesn't show any regressions: https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/629927

CyrusNajmabadi commented on src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LanguageServerExportProviderBuilder.cs:
```diff
@@ -0,0 +1,117 @@
+
+        // Load any Roslyn assemblies from the extension directory
+        using var _ = ArrayBuilder<string>.GetInstance(out var assemblyPathsBuilder);
+        assemblyPathsBuilder.AddRange(Directory.EnumerateFiles(baseDirectory, "Microsoft.CodeAnalysis*.dll"));
```
just doc IO logic.

ToddGrun replied:
Done!

CyrusNajmabadi commented on src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LanguageServerExportProviderBuilder.cs:
```diff
@@ -0,0 +1,117 @@
+    private readonly ILogger<ExportProviderBuilder> _logger;
+
+    // For testing purposes, track the last cache write task.
+    private static Task? s_cacheWriteTask;
```
```suggestion
    private static Task? s_cacheWriteTask_forTestingPurposesOnly;
```

CyrusNajmabadi commented on src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LanguageServerExportProviderBuilder.cs:
```diff
@@ -0,0 +1,117 @@
+    {
+        // Verify that we have exactly the MEF errors that we expect.  If we have less or more this needs to be updated to assert the expected behavior.
+        var expectedErrorPartsSet = new HashSet<string>(["CSharpMapCodeService", "PythiaSignatureHelpProvider", "CopilotSemanticSearchQueryExecutor"]);
+        var hasUnexpectedErroredParts = erroredParts.Any(part => !expectedErrorPartsSet.Contains(part));
```
could literally be a switch, vs a set.  but this is ok.

ToddGrun replied:
Seems a bit more concise in this form, but I don't feel strongly.

ToddGrun commented on src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LanguageServerExportProviderBuilder.cs:
```diff
@@ -0,0 +1,117 @@
+        // and not included in ExtensionAssemblyPaths (they get loaded into the default ALC).
+        // So manually add them to the MEF catalog here.
+        if (devKitDependencyPath != null)
+            assemblyPathsBuilder.AddRange(devKitDependencyPath);
```
Actually, this could just be an Add

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Remote/VisualStudioRemoteHostClientProvider.cs:
```diff
@@ -122,9 +131,10 @@ private VisualStudioRemoteHostClientProvider(
 
             // VS AsyncLazy does not currently support cancellation:
-            var client = await ServiceHubRemoteHostClient.CreateAsync(Services, configuration, _listenerProvider, serviceBroker, _callbackDispatchers, CancellationToken.None).ConfigureAwait(false);
+            var client = await ServiceHubRemoteHostClient.CreateAsync(Services, configuration, localSettingsDirectory, _listenerProvider, serviceBroker, _callbackDispatchers, CancellationToken.None).ConfigureAwait(false);
```
interesting.  we have an IThreadingContext, but we're not using the disposal token ehre.  feels... wrong.  though i see you're not changing that.

CyrusNajmabadi commented on src/VisualStudio/Core/Test.Next/Services/ServiceHubServicesTests.cs:
```diff
@@ -822,8 +822,8 @@ internal async Task TestSourceGenerationExecution_MinorVersionChange_NoActualCha
 
-        var remoteProcessId = await client.TryInvokeAsync<IRemoteProcessTelemetryService, int>(
-            (service, cancellationToken) => service.InitializeAsync(workspaceConfigurationService.Options, cancellationToken),
+        _ = await client.TryInvokeAsync<IRemoteInitializationService, (int, string)>(
```
would like names on these tuple parameters.

CyrusNajmabadi replied:
oh.  this is tests.  ignore.

CyrusNajmabadi commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -0,0 +1,198 @@
+        ThrowOnUnexpectedErrors(config, catalog);
+
+        // Try to cache the composition.
+        _ = WriteCompositionCacheAsync(compositionCacheFile, config, cancellationToken).ReportNonFatalErrorAsync();
```
is there any concern about two instances of VS loading, both calling into this, and both stomping on each other.  That's a general concern i have here unilaterally.  I'd actually like to talk in person to understand this.

CyrusNajmabadi commented on src/Workspaces/Remote/Core/ExportProviderBuilder.cs:
```diff
@@ -0,0 +1,198 @@
+        // Delete any existing cached files.
+        foreach (var fileInfo in directoryInfo.EnumerateFiles())
+        {
+            IOUtilities.PerformIO(fileInfo.Delete);
```
doc that you're itnentionally making it so that failing to delete isn't an issue.

CyrusNajmabadi commented on src/Workspaces/Remote/ServiceHub/Services/Initialization/RemoteInitializationService.cs:
```diff
@@ -0,0 +1,42 @@
+            => new RemoteInitializationService(arguments);
+    }
+
+    /// <summary>
+    /// Remote API.
+    /// </summary>
```
```suggestion
```

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Helpers/MefHostServicesHelpers.cs:
```diff
@@ -53,4 +52,23 @@ private static Assembly TryLoadNearbyAssembly(string assemblySimpleName)
+            ? Path.Combine(thisAssemblyFolder, assemblySimpleName + ".dll")
+            : null;
+
+        if (File.Exists(potentialAssemblyPath))
```
doc IO strategy.

ToddGrun replied:
I don't think File.Exists exposes IO failures
